### PR TITLE
Optional validation

### PR DESCRIPTION
**Pull request by xronos-i-am**
### Summary

This PR remove hardcoded inclusion of `ActiveModel::Validations` and allow to add it optionally. 

Component with validation [15 times slower](https://github.com/github/actionview-component/pull/192#issuecomment-580746645). I suppose it usable for 1 of 100 cases. Validation is too sloow to run it for every component

### CODE REVIEW

**Comment by jonspalmer** [comment]
I've run a variant benchmark myself and found slightly different results. If you simply don't use validation , but still call `validate` it's about 3 times slower. However, running a more realistic benchmark that actually includes rendering a view component and prove there was a difference between the two. I'm remain unconvinced that we need this at all.

```
class ErbComponent < ActionView::Component::Base
  validates :content, presence: true

  def initialize(message:)
    @message = message
  end
end

class ErbNoValidateComponent < ActionView::Component::Base
  def initialize(message:)
    @message = message
  end
end
```

and the benchmark run in the test suite:

```
  def test_benchmark
    Benchmark.ips do |x|
      x.report("with validation") do
        render_inline(ErbComponent, message: "bar") { "foo" }
      end
    
      x.report("without validation") do
        render_inline(ErbNoValidateComponent, message: "bar") { "foo" }
      end
    
      puts x.compare!
    end
  end
```

```
Warming up --------------------------------------
     with validation   759.000  i/100ms
  without validation   783.000  i/100ms
Calculating -------------------------------------
     with validation      8.021k (± 4.8%) i/s -     40.227k in   5.027458s
  without validation      8.045k (± 5.0%) i/s -     40.716k in   5.074260s

Comparison:
  without validation:     8045.3 i/s
     with validation:     8020.6 i/s - same-ish: difference falls within error
```

this issue is rendering dominates validate in terms of total time spent.

**Comment by jonspalmer** [comment]
If we want to support 'turning' off validation there are MUCH simpler ways to do it. Rather that the approach here just add a class_attribute

```ruby
class_attribute :skip_validation

def self.without_validation
  skip_validation = true
end

def self.with_validation
  skip_validation = false
end

....

validate! unless skip_validation

...
```

Them if you want to turnoff validation entirely just call `without_validation` in your ApplicationComponent and you're done (and could turn it back on in other components if you wanted)

**Comment by xronos-i-am** [comment]
> If we want to support 'turning' off validation there are MUCH simpler ways to do it. Rather that the approach here just add a class_attribute

keep in mind ability to change `ActionModel::Validations` to `dry-validation`

**Comment by dylnclrk** [comment]
For what it's worth, I prefer the on-by-default approach that @jonspalmer described. I wonder how many  users will want to make the performance tradeoff and turn them off. 

It seems better to have fewer hurdles to make use of validations.

**Comment by joelhawksley** [comment]
@xronos-i-am thanks for all of your work on this and the related PRs you've had in flight ❤️ 

In order to avoid introducing a breaking change, I'd prefer that we start by making it possible to disable validations, as well as to swap out the validation library. I think you're pretty close here in this PR.

**Comment by xronos-i-am** [comment]
@joelhawksley done

**Comment by jonspalmer** [comment]
I'm concerned that as coded this is still a breaking change that requires the inclusion of the validation module. I have a few other comments but no time for a little while. Will provide a more complete review later today.

File: lib/action_view/component/validatable.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+module ActionView
+  module Component # :nodoc:
+    module Validatable
+      extend ActiveSupport::Concern
+
+      included do
+        if ActionView::Component.extend_validations
+          include ActiveModel::Validations
+        end
```
Comment (jonspalmer) [inline lib/action_view/component/validatable.rb]: What is the importance of making this configurable? What's the cost of including this always even if someone is overriding `validate!`?

File: lib/action_view/component/validatable.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+        # method `validate!` actually performs validation and can be overrided
+        def validate_input!
+          return if self.class.validation_skiped?
+          validate!
+        end
+      end
```
Comment (jonspalmer) [inline lib/action_view/component/validatable.rb]: Why do we need the `validate_input!` wrapper vs just inlining in base as

```validate! unless skip_validation```

File: lib/action_view/component/validatable.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+      class_methods do
+        attr_accessor :skip_validation
+        attr_accessor :validation_decorator
```
Comment (jonspalmer) [inline lib/action_view/component/validatable.rb]: is `validation_decorator` unused?

File: test/action_view/component_test.rb [hunk 4]
```diff
@@ -365,6 +365,30 @@ def test_render_check
...
+  end
+
+  def test_dry_validation
+    assert_includes render_inline(DryValidationComponent, required: "text").text, "Rendered"
+
+    exception = assert_raises RuntimeError do
+      render_inline(DryValidationComponent)
+    end
+    assert_equal exception.message, "required: must be filled"
+  end
```
Comment (jonspalmer) [inline test/action_view/component_test.rb]: Is it important to test dry-validation here? The real test here is that "if I override `validate!` and throw exceptions those exceptions are not rescued` We could do that with any kind of override, there is nothing unique about the fact that it works with dry-validation (which would give us one less dev dependency)

File: test/app/components/without_validation_component.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class WithoutValidationComponent < ActionView::Component::Base
+  without_validation!
+
+  include ActiveModel::Validations
```
Comment (jonspalmer) [inline test/app/components/without_validation_component.rb]: do we need this `include`?

File: test/app/components/without_validation_component.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class WithoutValidationComponent < ActionView::Component::Base
```
Comment (jonspalmer) [inline test/app/components/without_validation_component.rb]: Its feels a little strange to have a `WithoutValidationComponent` that has a `validates` method. This isn't how you'd do it in your application and I worry about test code as misleading documentation. 

Instead how about we override the `validate!` method in this class and have it raise

```ruby
def validate!
  raise "validate! was called"
end
```

File: lib/action_view/component.rb [hunk 7]
```diff
@@ -9,14 +9,23 @@ module ActionView
...
 
+    mattr_accessor :extend_validations
+    @@extend_validations = true
+
+    def self.configure
+      yield self
+    end
+
     autoload :Base
+    autoload :Configuration
```
Comment (jonspalmer) [inline lib/action_view/component.rb]: Why is this needed as a result of these changes?

Comment (jonspalmer) [inline lib/action_view/component.rb]: What is this module?

File: README.md [hunk 8]
```diff
@@ -155,6 +155,48 @@ Which returns:
...
+      raise RuntimeError.new(result.errors.to_h)
+    end
+  end
+
+  attr_reader :email
+  private :email
+
+  def initialize(email: nil)
+    @email = email
+  end
```
Comment (jonspalmer) [inline README.md]: its just documentation code but the constructor needs to be public.

File: lib/action_view/component/validatable.rb (hunk 2 above)
Comment (xronos-i-am) [reply lib/action_view/component/validatable.rb]:
```ruby
include ActionView::Component::Validatable
validate_input!
```

vs 

```ruby
include ActionView::Component::Validatable
validate! unless skip_validation
```

In first case you have to know about only one method `validate_input!` of related `Validatable` module in `Base`. While the second one requires you to know about `skip_validation` also. That increase code coupling

File: lib/action_view/component/validatable.rb (hunk 3 above)
Comment (xronos-i-am) [reply lib/action_view/component/validatable.rb]: yes, mistake

File: test/action_view/component_test.rb (hunk 4 above)
Comment (xronos-i-am) [reply test/action_view/component_test.rb]: I think it is better to have real world example here

File: test/app/components/without_validation_component.rb (hunk 5 above)
Comment (xronos-i-am) [reply test/app/components/without_validation_component.rb]: no, mistake

File: test/app/components/without_validation_component.rb (hunk 6 above)
Comment (xronos-i-am) [reply test/app/components/without_validation_component.rb]: agree

File: lib/action_view/component.rb (hunk 7 above)
Comment (xronos-i-am) [reply lib/action_view/component.rb]: mistake

File: README.md (hunk 8 above)
Comment (xronos-i-am) [reply README.md]: `initialize` is public. The only private method here is `def email`. Do you want to make it public?

File: lib/action_view/component/validatable.rb (hunk 1 above)
Comment (xronos-i-am) [reply lib/action_view/component/validatable.rb]: I think it is not correct to load module that not used. It might not to has a big cost probably, but it is additional bytes/miliseconds that is unneccessary. People who prefer dry ecosystem and sequel make this choice because of performance as pointed by @rainerborene [here](https://github.com/github/actionview-component/pull/192#issuecomment-580766349) 

If we make optional validation mechanism, I think the next step could be to make `ActiveModel` optional at all

File: README.md (hunk 8 above)
Comment (jonspalmer) [reply README.md]: ok I misread the code

**Comment by jonspalmer** [comment]
Thanks for the detailed replies. I think we can simplify this entire setup by making the validation logic entirely swappable. How about just do this:

1. Introduce a new configuration setting, a string: `validation_module`. It defaults to "ActiveModel::Validations"
2. Add a new module ActionView::Component::NoOpValidations which has a single method `validate!` that does nothing.
3. in Base add something like the following code:

```ruby
if self.validation_module
  include self.validation_module.constantize
else
  include ActionView::Component::NoOpValidations
end
```

With that setup we are done and with very little code. Users would be free to do a few different things:

- turn off Validation entirely by setting `validation_module` to nil or false in config. 
- Add their own validation module. The contract would be that module HAS to provide a `validate!` method.

In either case users could later override the `validate!` method in any Component to adjust behavior.

Under this setup there is very much less code, and we don't need the indirection and confusion of wrappers like `validate_input` etc. Plus it's more obvious, easier to configure and more flexible.

Thoughts?

**Comment by xronos-i-am** [comment]
> Thoughts?

The problem is users will not be able to override `validation_module`:

```ruby
module AM
  def validate!
    puts "AM validate"
  end
end

module Noop
  def validate!
    puts "Noop"
  end
end

module Dry
  def validate!
    puts "Dry validate"
  end
end

class Base
  cattr_accessor :validation_module
  @@validation_module = 'AM'

  if self.validation_module
    include self.validation_module.constantize
  else
    include Noop
  end
end

class Child < Base
  self.validation_module = 'Dry'
end

Child.new.validate! # this outs "AM validate"
```

The reason is that value of `self.validation_module` of subclass resolves at the moment when subclass is inherited from `Base`. In that moment subclass itself is not evaluated yet. Related question on [SO](https://stackoverflow.com/questions/790626)

**Comment by jonspalmer** [comment]
Sure but I'm pretty sure this pattern would work:

```ruby
module AM
  def validate!
    puts "AM validate"
  end
end

module Noop
  def validate!
    puts "Noop"
  end
end

module Dry
  def validate!
    puts "Dry validate"
  end
end

class Base
  cattr_accessor :validation_module
  @@validation_module = 'AM'

  def self.inherited(subclass)
    subclass.include self.validation_module ? self.validation_module.constantize : Noop
  end

class Child < Base
 end

Child.new.validate! # this outs "AM validate"

Base.validation_module = 'Dry'

class Child2 < Base
end

Child2.new.validate! # this outs "Dry validate"
```

Provided the validation_module is set via Config before the subclasses are autoloaded this would give us what we need.

**Comment by xronos-i-am** [comment]
@jonspalmer done

File: lib/action_view/component/base.rb [hunk 9]
```diff
@@ -14,6 +13,13 @@ class Base < ActionView::Base
       class_attribute :content_areas, default: []
       self.content_areas = [] # default doesn't work until Rails 5.2
 
+      # `validation_module` should be set to a string if present
+      class_attribute :validation_module, default: "ActiveModel::Validations"
+      self.validation_module = "ActiveModel::Validations" # default doesn't work until Rails 5.2
+
+      # when set to `true` validation does not run
+      class_attribute :skip_validation
```
Comment (rainerborene) [inline lib/action_view/component/base.rb]: I think we should deprecate the usage of `skip_validation: false`, meaning that validations will be gone on the next major version of `actionview-component`. Other than that, this PR is 💯

Comment (jonspalmer) [reply lib/action_view/component/base.rb]: I think we just eliminate it entirely. I don't think we need to introduce this concept at all.

File: test/app/components/dry_validation_component.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  def initialize(required: nil)
+    @required = required
+  end
+
+  private def component_params
+    {required: required}
+  end
+end
+
+ActionView::Component::Base.validation_module = "ActiveModel::Validations"
```
Comment (jonspalmer) [inline test/app/components/dry_validation_component.rb]: Can we use `ActiveModel::Validations.name` here?

File: test/app/components/dry_validation_component.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+module DryValidations
+  def validate!
+    result = self.class.contract.call(component_params)
+
+    unless result.errors.empty?
+      full_messages = result.errors.to_h.map { |f, messages| "#{f}: #{messages.join(', ')}" }
+      raise RuntimeError.new(full_messages.join(", "))
+    end
+  end
+end
```
Comment (jonspalmer) [inline test/app/components/dry_validation_component.rb]: I understand the 'real world' example argument but we're trying to test/prove that you can provide an alternative `validate!` method. This feels like a few more moving parts that we need to do that. Can we simplify (maybe not have to need to use dry-validation and force the reader to understand that gem to grok this test). 

I'm all for documenting using dry-validation in the Readme as a real example.

File: README.md [hunk 12]
```diff
@@ -155,6 +155,35 @@ Which returns:
...
+```
+
+You can also skip validation in some components with `skip_validation`:
+
+`app/components/validation_disabled.rb`:
+```ruby
+class ValidationDisabled < ActionView::Component::Base
+  self.skip_validation = true
+end
+```
```
Comment (jonspalmer) [inline README.md]: I don't see the need for both mechanisms and its confusing to provide the choice. Most folks will stick wit h the built in ActiveModel Validations. If you want finer control you supply your own module. Users are free to build features like `skip_validation` into their own validation module but I don't think we need it out of the box.

**Comment by joelhawksley** [comment]
@xronos-i-am thank you so much for your work on this PR, and to @jonspalmer for the thorough review.

After speaking with @rainerborene, @jonspalmer, and several folks here at work, we think it's best to move forward with removing validations from the library in `v2`. 

We'll start by moving the call to `validate!` to a `before_render_check` method, which can be subclassed to be a no-op.

Then, when we introduce the new `ViewComponent::Base` class, we will remove the `validate!` call and the inclusion of `ActiveModel::Validations`.

**Comment by xronos-i-am** [comment]
@joelhawksley ok. When the second version is planned to delivery?

**Comment by joelhawksley** [comment]
@xronos-i-am before RailsConf / 6.1 😄

**Comment by xronos-i-am** [comment]
@joelhawksley ok ))
